056-zk-kraft-migration.md

```diff
@@ -0,0 +1,451 @@
+
+One of the main points to highlight is that such a process doesn't support migration to a KRaft-based cluster running in "combined" mode:
+
+* Cannot migrate from a ZooKeeper based configuration to a cluster with nodes hosting both controller and broker (so actually reducing the number of nodes by removing the ZooKeeper ones).
+* The migration works by having ZooKeeper nodes to be replaced by new KRaft controller nodes.
+* Supporting "combined" mode is actually one of the rejected [alternatives](https://cwiki.apache.org/confluence/display/KAFKA/KIP-866+ZooKeeper+to+KRaft+Migration#KIP866ZooKeepertoKRaftMigration-CombinedModeMigrationSupport) in KIP-866.
+
+It means that the user environment should be sized to host more nodes running in parallel, because during the migration process both ZooKeeper nodes and KRaft controllers will be running.
+The end of the migration will involve shutting down the ZooKeeper nodes reducing the resources consumption.
+
+## Motivation
+
+The KRaft mode has been declared to be production-ready since Apache Kafka 3.3 release by the Apache Kafka project. 
+Anyway, there are still missing features as reported in the official [documentation](https://kafka.apache.org/documentation/#kraft_missing).
+The ZooKeeper mode has been deprecated after the Apache Kafka 3.5 version was released while providing the migration process to KRaft mode as preview.
+The coming Apache Kafka 3.6 release will have such a migration process supported as GA and the ZooKeeper mode will be supported until Apache Kafka 3.7 (January 2024) but removed after that.
+Starting from Apache Kafka 4.0 (April 2024) only KRaft mode will be supported.
+
+Based on the above Apache Kafka community plan, more details in [KIP-833](https://cwiki.apache.org/confluence/display/KAFKA/KIP-833%3A+Mark+KRaft+as+Production+Ready), the Strimzi project needs to be prepared to allow users to migrate their current clusters from running in ZooKeeper mode to KRaft mode.
+
+## Proposal
+
+### Prerequisites
+
+In order to start the ZooKeeper to KRaft migration, the following prerequisites should be met:
+
+* A Strimzi-managed Apache Kafka cluster running in ZooKeeper mode.
+* Running at Kafka 3.5.0 or higher and with `inter.broker.protocol.version` set to 3.5 or higher.
+* Configured to use one or more node pools, with `KafkaNodePool` custom resources, all with the `broker` role.
+* No existing node pool(s) which include the `controller` role.
+* The Unidirectional Topic Operator has to be configured, or not used at all, as the previous Bidirectional doesn't work in KRaft mode.
+* Either KRaft mode is opted into (using the `+UseKRaft` feature gate), or the `+UseKRaft` feature gate has reached beta level and defaults to enabled.
```

Comment:
```suggestion
* The `UseKRaft` feature gate should be enabled.
```

Reply:
This was suggested by @tombentley ... I would like to have his opinion on this change.

Reply:
Well, this says the same without the need to think in conditionals (enabled by default is still enabled). So I assume this is better ;-).

Reply:
How about this, which is less wordy than the original, but more explicit than Jakub's suggestion (which assumes the reader has knowledge of feature gate progression).

```suggestion
* The `UseKRaft` feature gate should be enabled (explictly, or via it being beta status).
```

Reply:
Yeah I think being more explicit helps readers.